editorial pass on section 3 (path management)

mirjak opened the pull request:
Please have a careful review as there have been some thing were old text was redundant or needed to be adopted/clarified based on newer text

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -379,25 +379,26 @@ should not cause linkability issue.
-
-When the multipath extension is negotiated, a client that wants to use an
+To open a new path, an endpoint MUST use a new connection ID associated
+with unused Path ID.
```
```suggestion
with an unused Path ID.
```

huitema replied:
Please also fix the upper case "T" in "To open"

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -379,25 +379,26 @@ should not cause linkability issue.
+To open a new path, an endpoint MUST use a new connection ID associated
+with unused Path ID.
+When sending a PATH_RESPONSE frame, an endpoint MUST use a connection ID associated to
+the same Path ID as used in packet that contained the PATH_CHALLENGE frame.
```
```suggestion
the same Path ID as used in the packet that contained the PATH_CHALLENGE frame.
```

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -412,16 +413,16 @@ validated, the anti-amplification limit from {{Section 8 of QUIC-TRANSPORT}}
 An endpoint that receives packets on a new path and does not want to establish
 this path is expected to close the path by sending a PATH_ABANDON
-on another path, as specified in section {{path-close}}.
+on another path, as specified in Section {{path-close}}.
```
```suggestion
on another path, as specified in {{path-close}}.
```

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -430,25 +431,22 @@ Long delays may cause the peer to repeat the PATH_CHALLENGE and eventually
+Any frames can be sent on any open path.
 If the PATH_ACK is preferred to be sent on the same path as the acknowledged
-packet (see {{compute-rtt}} for further guidance), it can be beneficial
+packet (see Section {{compute-rtt}} for further guidance), it can be beneficial
```
```suggestion
packet (see {{compute-rtt}} for further guidance), it can be beneficial
```

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -430,25 +431,22 @@ Long delays may cause the peer to repeat the PATH_CHALLENGE and eventually
-{{path-close}}.
+As the used Path ID is anyway consumed,
+the endpoint MUST explicitly close the path, as specified in
+Section {{path-close}}.
```
```suggestion
{{path-close}}.
```

huitema replied:
+1

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -522,31 +514,46 @@ for subsequent connections, it might be difficult for the client
+(see {{Section 5.1.1 of QUIC-TRANSPORT}}), at least for all paths with a Path ID other than 0.
+Each Path ID has its own connection ID sequence number space whose initial value is 0.
+
+Endpoints can also use PATH_NEW_CONNECTION_ID and
+PATH_RETIRE_CONNECTION_ID for the initial path with Path ID 0,
+however, the use of NEW_CONNECTION_ID and RETIRE_CONNECTION_ID
+is still valid and endpoints need to process these frames accordingly
+as corresponding to Path ID 0.
```
Note regarding other editorial passes: we should then remove redundant text in the frame section. We may keep this text here, but I think merging it with Section 2.3 would avoid repeating this for all the PATH_* frames.

mirjak replied:
I don't think the Frames section says this...? Which section do you mean by 2.3?

mirjak replied:
@qdeconinck what do you mean?

Yanmei-Liu replied:
I guess @qdeconinck is referring to [Section 7.4](https://datatracker.ietf.org/doc/html/draft-ietf-quic-multipath-13#name-path_new_connection_id-fram):

> Note that the NEW_CONNECTION_ID frame can only be used to issue or retire connection IDs for the initial path with Path ID 0.

qdeconinck replied:
Yeah, that was what @Yanmei-Liu said. What I meant is that we should say at one place that NEW_CONNECTION_ID, RETIRE_CONNECTION_ID and ACK frames are still valid and are equivalent to their PATH_* versions with Path ID 0. We can delegate this to a further issue, but we can just merge this for now.

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -522,31 +514,46 @@ for subsequent connections, it might be difficult for the client
-connection ID sequence number space whose initial value is 0. On non-initial paths
-(i.e., Path ID different from 0), the initial path connection ID is provided by a
-PATH_NEW_CONNECTION_ID frame with a sequence number value of 0.
+If for any reasons an endpoints does not want to issue connection IDs for all
```
```suggestion
If for any reason an endpoint does not want to issue connection IDs for all
```

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -522,31 +514,46 @@ for subsequent connections, it might be difficult for the client
+
+Similarly, endpoints SHOULD consume Path IDs in a continuous way, i.e., when
+creating paths. However, endpoints cannot expect to receive new connection IDs
+or path initiation attempts with in order use of Path IDs
```
```suggestion
or path initiation attempts with in-order use of Path IDs
```

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -558,70 +565,47 @@ cannot be issued more than once on the same connection
 or the value of initial_max_path_id transport parameter if no MAX_PATH_ID frame was received yet.
 Receipt of a frame with a greater Path ID is a connection error as specified
-in {{frames}}.
+in Section {{frames}}.
```
```suggestion
in {{frames}}.
```

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -558,70 +565,47 @@ cannot be issued more than once on the same connection
-(see Section {{paths-and-cids-blocked-frame}}) to inform the peer that a new path
-identifier was needed but the current limit set by the peer prevented the
-creation of the new path.
+(see Section {{paths-and-cids-blocked-frame}}) to inform the peer that
```
```suggestion
(see {{paths-and-cids-blocked-frame}}) to inform the peer that
```

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -558,70 +565,47 @@ cannot be issued more than once on the same connection
-however, the use of NEW_CONNECTION_ID and RETIRE_CONNECTION_ID
-is still valid as well and endpoints need to process these frames accordingly
-as corresponding to Path ID 0.
+An endpoints is supposed to retire connection ID that are not used anymore,
```
```suggestion
An endpoint is supposed to retire connection ID that are not used anymore,
```

huitema replied:
I would say, "An endpoint is expected to retire connection IDs that are not used anymore,"

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -558,70 +565,47 @@ cannot be issued more than once on the same connection
+An endpoints is supposed to retire connection ID that are not used anymore,
+and the server is supposed to provide
+replacements, as specified in {{Section 5.1.2. of QUIC-TRANSPORT}}.
+As such when receiving a PATH_RETIRE_CONNECTION_ID frame, an endpoint
```
```suggestion
As such, when receiving a PATH_RETIRE_CONNECTION_ID frame, an endpoint
```

huitema replied:
Might be better to just remove "As such,"

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -558,70 +565,47 @@ cannot be issued more than once on the same connection
 for all active paths using the PATH_NEW_CONNECTION_ID after retirement
 of an old connection ID, an endpoint MAY send
-a PATH_CIDS_BLOCKED (see {{paths-and-cids-blocked-frame}})
+a PATH_CIDS_BLOCKED (see Section {{paths-and-cids-blocked-frame}})
```
```suggestion
a PATH_CIDS_BLOCKED (see {{paths-and-cids-blocked-frame}})
```

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -558,70 +565,47 @@ cannot be issued more than once on the same connection
 Retirement of connection IDs will not retire the Path ID
 that corresponds to the connection ID or any other path resources
-as the packet number space is associated with a path.
+as the packet number space is associated with to the Path ID.
```
```suggestion
as the packet number space is associated to the Path ID.
```

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -642,76 +626,75 @@ indicated as "standby" by the peer or non-usage of some locally available paths.
 An endpoint that detects a path breakage can also explicitly close the path
-by sending a PATH_ABANDON frame (see {{path-close}}) in order to avoid
-that its peer keeps using it and enable faster switch over to a standby path.
+by sending a PATH_ABANDON frame (see Setion {{path-close}}) in order to avoid
```
```suggestion
by sending a PATH_ABANDON frame (see {{path-close}}) in order to avoid
```

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -642,76 +626,75 @@ indicated as "standby" by the peer or non-usage of some locally available paths.
-from implicit signals such as idle time or packet losses
-(see {{idle-time-close}}) or for any other reason, such as management
-of local resources.
+does not indicate path abandonment (see further Section {{consume-retire-cid}}).
```
```suggestion
does not indicate path abandonment (see further {{consume-retire-cid}}).
```

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -642,76 +626,75 @@ indicated as "standby" by the peer or non-usage of some locally available paths.
+does not indicate path abandonment (see further Section {{consume-retire-cid}}).
+
+PATH_ABANDON frames can be sent on any path,
+not only the path that is intended to be closed.
```
```suggestion
not only on the path that is intended to be closed.
```

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -642,76 +626,75 @@ indicated as "standby" by the peer or non-usage of some locally available paths.
+of the number space associated to the path SHOULD be retained for
+3 PTO after the PATH_ABANDON frame has been received.
+This avoids generating spurious stateless packets, as discussed in
+{{spurious-stateless-reset}}, and helps to acknowledge any
```
```suggestion
{{spurious-stateless-reset}}, and helps acknowledge any
```

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -642,76 +626,75 @@ indicated as "standby" by the peer or non-usage of some locally available paths.
+3 PTO after the PATH_ABANDON frame has been received.
+This avoids generating spurious stateless packets, as discussed in
+{{spurious-stateless-reset}}, and helps to acknowledge any
+potentially reordered, outstanding packets from the peer (see Section {ack-after-abandon}).
```
```suggestion
potentially reordered, outstanding packets from the peer (see {{ack-after-abandon}}).
```

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -642,76 +626,75 @@ indicated as "standby" by the peer or non-usage of some locally available paths.
 tries to initiate a path and the path cannot be established, it will send a
-PATH_ABANDON frame (see {{path-initiation}}). An endpoint may also decide
-to abandon a path for any reason, for example, removing a hole from
+PATH_ABANDON frame (see Section {{path-initiation}}). An endpoint may also decide
```
```suggestion
PATH_ABANDON frame (see {{path-initiation}}). An endpoint may also decide
```

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -720,36 +703,26 @@ willing to process the issued connection IDs for the abandoned path,
-frame for the last open path, it MAY instead try to open a new path, if
+and enter the closing state. Alternatively, a client MAY instead try to open a new path, if
 available, and only initiate connection closure if path validation fails
 or a CONNECTION_CLOSE frame is received from the server. Similarly
```
```suggestion
or a CONNECTION_CLOSE frame is received from the server. Similarly,
```

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -775,40 +748,38 @@ PATH_ABANDON frame.
+Due to network delays, packets sent on an abandoned path can
+arrive well after the connection IDs have been retired.
+If not recognized as bound to the local
+connection, such packet trigger the peer to send a Stateless Reset
```
```suggestion
connection, such packet triggers the peer to send a Stateless Reset
```

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -775,40 +748,38 @@ PATH_ABANDON frame.
+When an endpoint sends a PATH_ABANDON frame, there may
+still be some packets in transit from the peer.
+Further, if an endpoint receives a PATH_ABANDON frame, it may still receive
+reordered packets on the abanded path. Endpoints SHOULD
```
```suggestion
reordered packets on the abandoned path. Endpoints SHOULD
```

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -775,40 +748,38 @@ PATH_ABANDON frame.
+still be some packets in transit from the peer.
+Further, if an endpoint receives a PATH_ABANDON frame, it may still receive
+reordered packets on the abanded path. Endpoints SHOULD
+promptly send PATH_ACK frames for all unacknowledge packets received on
```
```suggestion
promptly send PATH_ACK frames for all unacknowledged packets received on
```

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -775,40 +748,38 @@ PATH_ABANDON frame.
+Further, if an endpoint receives a PATH_ABANDON frame, it may still receive
+reordered packets on the abanded path. Endpoints SHOULD
+promptly send PATH_ACK frames for all unacknowledge packets received on
+an abondoned path if path state is still retained to do so.
```
```suggestion
an abandoned path if path state is still retained to do so.
```

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -775,40 +748,38 @@ PATH_ABANDON frame.
+When an endpoint finally deletes all state associated with the path,
+the packets sent over the path and not yet acknowledged MUST be considered lost.
+PATH_ACK frames received with an abandoned Path ID are silently ignored,
+as specified in Section {{frames}}.
```
```suggestion
as specified in {{frames}}.
```

qdeconinck reviewed (commented):
Overall good, many small fixes.

huitema commented on draft-ietf-quic-multipath.md:
```diff
@@ -379,25 +379,26 @@ should not cause linkability issue.
+multipath support, endpoints can start using multiple paths when both endpoints
+have issued available connection IDs for at least one unused, common Path ID,
+as the same Path ID is used in both directions.
+
```
Not sure that this rewrite is better than the original. You could just strike out the last line, "This document does not discuss when a client decides to initiate a new path."

mirjak replied:
The initial text is not correct. 

"After completing the handshake, endpoints have agreed to enable
multipath support." -> only if both ends indicated multipath support (but I don't think we need to repeat this here.

I think this happened because this used to be use sentence we split it up (wrongly) at some point.

Please see another proposal to split this into two sentence above.

huitema replied:
New proposal works.

huitema commented on draft-ietf-quic-multipath.md:
```diff
@@ -379,25 +379,26 @@ should not cause linkability issue.
+have issued available connection IDs for at least one unused, common Path ID,
+as the same Path ID is used in both directions.
+
+This documents specfies path initiation (see Section {{path-initiation}}),
+issuing and retirement of per-path connection IDs (see Section
+{{consume-retire-cid}}), path status management (see Section {{path-state}}) and
+path closure (see Section {{path-close}}).
```
+1

huitema commented on draft-ietf-quic-multipath.md:
```diff
@@ -379,25 +379,26 @@ should not cause linkability issue.
+path closure (see Section {{path-close}}).
+However, this document does not specify when a client decides to initiate or close a path,
+or how multiple active paths are used for packet sending.
 
```
"are used for sending packets." ?

huitema commented on draft-ietf-quic-multipath.md:
```diff
@@ -430,25 +431,22 @@ Long delays may cause the peer to repeat the PATH_CHALLENGE and eventually
 Section {{path-close}} apply.
 
-PATH_ACK frames (defined in {{mp-ack-frame}}) can be returned on any path.
+Any frames can be sent on any open path.
```
Previous version is better. PATH_CHALLENGE and PATH_RESPONSE frames cannot be "sent on any open path".

mirjak replied:
The point is PATH_CHALLENGE and PATH_RESPONSE frames are not special. There are no restrictions for any frame on which path to sent. We don't say this anywhere. I think it good to say that once somewhere.

huitema replied:
They are special. You cannot send a challenge or a response for path 1 on path2...

huitema replied:
I think we should keep the old text, for the same reason that we have text stating that "PATH_ABANDON frames can be sent on any path, not only on the path that is intended to be closed." It is important to assert that loudly, so implementers do not start assuming that PATH_ACK can only be sent on the path on which the acked packets were sent.

mirjak replied:
Okay you are right. Reverting it.

mirjak replied:
```suggestion
PATH_ACK frames (see {{mp-ack-frame}}) can be returned on any path.
```

huitema commented on draft-ietf-quic-multipath.md:
```diff
@@ -522,31 +514,46 @@ for subsequent connections, it might be difficult for the client
+On non-initial paths
+(i.e., Path ID different from 0), the initial path connection ID is provided by a
+PATH_NEW_CONNECTION_ID frame with a sequence number value of 0.
 In order to let the peer open new paths, it is RECOMMENDED to proactively
```
No. Previous text is better. You may want to say that it is OK to send legacy NEW_CONNECTION_ID on path 0, but your text ends up saying that NEW_CONNECTION_ID frames cannot be used on path with number not equal to zero,
which is not what we agreed. How about:

The PATH_NEW_CONNECTION_ID frame is used to issue new connection IDs for any path. The NEW_CONNECTION_ID frame MAY be used to provide new connection IDs for path number 0. Connection ID sequence numbers start at 0.

mirjak replied:
This is not a meant as a replacement. I simply moved this sentence up (see removed lines 547, 548, and 579 below to find the original sentence). And I removed the first sentence because I found it redundant with the text above (see lines 523-527) and also text in the frames section.

However, we can maybe also just remove this sentence as well as we say in line 528 already this:
"Each Path ID has its own connection ID sequence number space whose initial value is 0."

huitema replied:
OK, let's remove the sentence.

mirjak replied:
done

huitema commented on draft-ietf-quic-multipath.md:
```diff
@@ -642,76 +626,75 @@ indicated as "standby" by the peer or non-usage of some locally available paths.
-CONNECTION_CLOSE ({{Section 10.2 of QUIC-TRANSPORT}}) or a Stateless
-Reset ({{Section 10.3 of QUIC-TRANSPORT}}) closes the connection.
+At any time in the connection, each endpoint can decide to
+abandon a paths, for example following changes in local
```
"abandon a path," (singular)

mirjak replied:
```suggestion
abandon a path, for example following changes in local
```

huitema commented on draft-ietf-quic-multipath.md:
```diff
@@ -642,76 +626,75 @@ indicated as "standby" by the peer or non-usage of some locally available paths.
+of local resources.
 
-An endpoint that wants to close a path MUST explicitly
-terminate the path by sending a PATH_ABANDON frame (see {{path-abandon-frame}}).
```
Can we keep that text? It says something very clearly -- that there is no other way to close a path. That's the main message of the "close path" section.

huitema reviewed (commented):
There are lots of changes here, probably a bit too much. I flagged a few places were changes should reverted or rewritten.

mirjak commented on draft-ietf-quic-multipath.md:
```diff
@@ -379,25 +379,26 @@ should not cause linkability issue.
-multipath support. They can start using multiple paths when both endpoints
-have issued available connection IDs for at least one unused Path ID.
-This document does not discuss when a client decides to initiate a new path.
+After completing the handshake indicating
+multipath support, endpoints can start using multiple paths when both endpoints
+have issued available connection IDs for at least one unused, common Path ID,
+as the same Path ID is used in both directions.
```
```suggestion
After completing the handshake indicating
multipath support, endpoints can start using multiple paths.
An endpoint can open a new path when both endpoints
have issued available connection IDs for at least one unused, common Path ID,
as the same Path ID is used in both directions.
```
Is this better?

huitema replied:
Yes.

mirjak commented on draft-ietf-quic-multipath.md:
```diff
@@ -379,25 +379,26 @@ should not cause linkability issue.
-it MUST pick a Connection ID with the same Path ID for sending the PATH_RESPONSE.
-
-When the multipath extension is negotiated, a client that wants to use an
+To open a new path, an endpoint MUST use a new connection ID associated
```
```suggestion
to open a new path, an endpoint MUST use a new connection ID associated
```

mirjak commented on draft-ietf-quic-multipath.md:
```diff
@@ -379,25 +379,26 @@ should not cause linkability issue.
+{{consume-retire-cid}}), path status management (see {{path-state}}) and
+path closure (see {{path-close}}).
+However, this document does not specify when a client decides to initiate or close a path,
+or how multiple active paths are used for packet sending.
```
```suggestion
or how multiple active paths are used for sending.
```

mirjak commented:
@huitema yes there are quite some changes and especially I moved text around to avoid redundancy which makes it really hard to read the diff. I don't think I removed anything really so. Please check my replies to your comments! Thanks!

mirjak commented on draft-ietf-quic-multipath.md:
```diff
@@ -522,31 +515,46 @@ for subsequent connections, it might be difficult for the client
 ### Issuing New Connection IDs
 
-The PATH_NEW_CONNECTION_ID frame is used to issue new connection IDs for all paths.
+On non-initial paths
+(i.e., Path ID different from 0), the initial path connection ID is provided by a
+PATH_NEW_CONNECTION_ID frame with a sequence number value of 0.
```
```suggestion
```

Yanmei-Liu commented on draft-ietf-quic-multipath.md:
```diff
@@ -412,16 +414,16 @@ validated, the anti-amplification limit from {{Section 8 of QUIC-TRANSPORT}}
-contain a path challenge. Such packets are valid if they can be properly decrypted
-and if they contain a valid connection ID.
+contain a PATH_CHALLENGE frame. Such packets are valid if they can be properly decrypted
+based on a valid connection ID.
```
The new sentence is not correct: AEAD decryption is not based on connection ID, but based on a Path ID as part of nonce. The original text is correct on logic, but hard to read. I'd suggest the new text changes to:

```suggestion
and contain a valid Connection ID.
```

qdeconinck replied:
The "based on" does not bother me, but the "given" version is maybe less ambiguous -- both of these versions works. I would not take the "and contain" as it would alter the original message of this sentence.

Yanmei-Liu reviewed (approved):
Looks good.
